src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -628,6 +628,49 @@ public bool Remove(TKey key)
             return false;
         }
 
+        public bool Remove(TKey key, out TValue value)
+        {
+            if (key == null)
+            {
+                value = default(TValue);
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
+            }
+
+            if (buckets != null)
```

Comment:
@jkotas thanks for the elaborate answer. I helps me understand more internals. I had not given a thought that the `operator []` would do bounds checking... took it for plain simple C / C++ array access.